Automattic/wp-calypso
# Checkout: Switch Web Payments to Stripe

The current Web Payment system (eg: Apple Pay) has special logic for creating PaymentRequest objects, showing a payment form, and sending that data to the server for a special checkout flow. 

As part of replacing Paygate usage in Calypso, this PR swaps out most of the custom code for the Stripe Elements _Payment Request Button_. This button has its own Payment Request object that abstracts over Apple Pay, Google Pay, etc, and integrates tightly with our other Stripe Elements payment flow (added in #34469) to create a Stripe Payment Method. It then sends that token to the server in the same manner as the current credit card flow.

![apple-pay-form](https://user-images.githubusercontent.com/2036909/63536903-8f756800-c4e2-11e9-89f9-e4eeae20ed31.png)

#### Testing instructions

1. Sandbox the store.
2. Use a proxy to forward requests from `https://wordpress.com` to localhost. This can be a complex step. Let me know if you have any questions about doing this. It's necessary because the button will only appear on a secure connection and on a verified url.
3. Make sure you have a real credit card added to Apple Pay in your local Safari. Look online for how to do this if you don't have it set up already or let me know and I can help.
3. Add a plan to your cart. Make sure you see the "dev" logo in the lower right of the screen.
4. Visit the checkout page in Safari. Apple Pay exists only in Safari on Mac OS.
5. Click the Apple Pay tab above the checkout form.
6. Click the Pay button.
7. When the Apple Pay sheet appears, authorize the payment. Don't worry, you won't be charged as long as you are sandboxing the store.
8. Verify that the transaction goes through correctly and that you are redirected to the next step of the checkout process (which may be a special offer).

### To do

- [x] Take care of updates to paymentRequest: https://github.com/stripe/react-stripe-elements/issues/284
- [x] Test errors and failures
- [x] Try to abstract away all the memoizations
- [x] Make sure country/postal are passed correctly
- [x] Verify purchase record (receipt, ownership, etc.) is correct on the server
- [x] Consider renewal flows
- [x] Convert amounts to correct values: D31850-code
- [x] Add better loading indicator for button
- [x] Disable button when country/postal are empty
- [x] Verify non-US currencies are shown and work
